service/matching/taskQueueManager.go

```diff
@@ -187,7 +193,8 @@ func newTaskQueueManager(
 		config:              taskQueueConfig,
 		pollerHistory:       newPollerHistory(),
 		outstandingPollsMap: make(map[string]context.CancelFunc),
-		signalFatalProblem:  e.unloadTaskQueue,
+		clock:               clock.NewRealTimeSource(),
+		signalFatalProblem:  e.requestUnload,
```

Comment:
This name signalFatalProblem bugs me so much. It is only used to unload task queue at the moment. Can we rename this signalFatalProblem to just unloadTaskQueue? Can always change it to something different in future when it become something different other than just unload task queue.

Reply:
The difference is in which entity decides a course of action. My suggestion is that matchingEngineImpl is the policy entity and that taskQueueManagerImpl is managed. Thus taskQueueManagerImpl needs a way to say "I'm busted" and the higher level entity (i.e. matchingEngineImpl) decides what to do with that. Maybe nothing, maybe Stop() followed by Start(), maybe a new instance, maybe just delete, maybe Start() on a different host. The point is to keep the managed entity ignorant of what the implementation is going to do so that these implementations can be decoupled.

Reply:
Understand your intention. It just feel the code become harder to read and I would like not to pay that price until it become a clear need.